### PR TITLE
Add setting CSS property with important priority

### PR DESCRIPTION
**Pull request by maximzasorin**
Hello, I added the ability to set properties with an important priority. No we can use it as follows:

```js
$('#testFragment').css('paddingRight', '10px !important');
```

In the tests, I did not use `getComputedStyle` because of an error in jsdom, for which inline styles are always in priority, see: https://github.com/jsdom/jsdom/issues/2485

#169

### CODE REVIEW

**Comment by maximzasorin** [comment]
@webpro Please comment on PR.

**Comment by webpro** [comment]
@maximzasorin My sincere apologies, I've totally missed it! Greenkeeper was polluting this space with PRs. Looking into it right now.

File: src/css.js [hunk 1]
```diff
@@ -46,16 +44,21 @@ export const css = function(key, value) {
...
   each(this, element => {
     for(prop in styleProps) {
+      let important = undefined;
+      if(typeof styleProps[prop] === 'string' && styleProps[prop].match(/!important$/)) {
+        styleProps[prop] = styleProps[prop].replace(/\s*!important$/, '');
+        important = 'important';
+      }
       if(styleProps[prop] || styleProps[prop] === 0) {
-        element.style[prop] = styleProps[prop];
+        element.style.setProperty(prop, styleProps[prop], important);
```
Comment (webpro) [inline src/css.js]: Let's optimize by doing something like this (not tested):

```
const important = /!important$/.test(styleProps[prop]) ? 'important' : undefined;
element.style.setProperty(prop, styleProps[prop].replace(/!important$/, ''), important);
```

Then the logic above it isn't executed if we're removing a property.

**Comment by webpro** [comment]
Btw, why don't we use this API:

```
$('#testFragment').css('paddingRight', '10px', 'important');
```

Makes implementation faster and simpler. How does jQuery do this?

**Comment by maximzasorin** [comment]
> Btw, why don't we use this API:
> ```
> $('#testFragment').css('paddingRight', '10px', 'important');
> ```
What will the API look like when setting multiple properties?

> How does jQuery do this?

As far as I know jQuery does not support important properties via [css](http://api.jquery.com/css/) method. From docs:

> Note: .css() ignores !important declarations. So, the statement $( "p" ).css( "color", "red !important" ) does not turn the color of all paragraphs in the page to red. It's strongly advised to use classes instead; otherwise use a jQuery plugin.

But they have an issue: https://github.com/jquery/jquery/issues/3713

File: src/css.js (hunk 1 above)
Comment (maximzasorin) [reply src/css.js]: Thanks, I'll fix it.

**Comment by maximzasorin** [comment]
@webpro Added fix.

File: src/css.js [hunk 2]
```diff
@@ -46,16 +44,20 @@ export const css = function(key, value) {
...
   }
 
   each(this, element => {
     for(prop in styleProps) {
       if(styleProps[prop] || styleProps[prop] === 0) {
-        element.style[prop] = styleProps[prop];
+        const important = /!important$/.test(styleProps[prop]) ? 'important' : undefined;
+        if(typeof styleProps[prop] === 'string') {
+          styleProps[prop] = styleProps[prop].replace(/\s*!important$/, '');
+        }
```
Comment (webpro) [inline src/css.js]: Sorry for being picky, how about this (one line and slightly better readable?):

```
styleProps[prop] = important ? styleProps[prop].replace(/\s*!important$/, '') : styleProps[prop];
```

File: test/spec/css.js [hunk 3]
```diff
@@ -51,6 +51,48 @@ describe('css', function() {
...
+        element[0].style.getPropertyValue('font-style')
+      ];
+      var actualPriority = [
+        element[0].style.getPropertyPriority('font-weight'),
+        element[0].style.getPropertyPriority('font-style')
+      ];
+      assert(actualValue[0] === expected[0]);
+      assert(actualPriority[0] === 'important');
+      assert(actualValue[1] === expected[1]);
+      assert(actualPriority[1] === '');
```
Comment (webpro) [inline test/spec/css.js]: Thanks for adding the tests, super useful! I know I have this `actual`/`expected` style everywhere, but could you please consider this for conciseness here:

```
assert(element[0].style.getPropertyValue('font-weight') === expected[0]);
assert(element[0].style.getPropertyPriority('font-weight') === 'important');
assert(element[0].style.getPropertyValue('font-style') === expected[1]);
assert(element[0].style.getPropertyPriority('font-style') === '');
```

File: test/spec/css.js [hunk 4]
```diff
@@ -51,6 +51,48 @@ describe('css', function() {
       assert.doesNotThrow(fn, TypeError);
     });
 
+    it('should set style property with important priority on elements', function() {
+      var expected = '15px';
+      $('#testFragment').css('paddingRight', expected + ' !important');
+      var element = $('#testFragment')[0];
+      var actualValue = element.style.getPropertyValue('padding-right');
+      var actualPriority = element.style.getPropertyPriority('padding-right');
+      console.log('actual/expected', actualValue, expected);
```
Comment (webpro) [inline test/spec/css.js]: Please remove this line

**Comment by webpro** [comment]
When sending the tests to BrowserStack, the CSS tests fail in IE and Edge:

* Windows 8, Internet Explorer 10.0
* Windows 7, Internet Explorer 11.0
* Windows 10, Edge 17.0
* Windows 10, Edge 18.0

Without any useful error/trace (`Unspecified AssertionError`). Do you happen to have one of these ready to test/debug?

File: test/spec/css.js (hunk 4 above)
Comment (maximzasorin) [reply test/spec/css.js]: Removed.

File: test/spec/css.js (hunk 3 above)
Comment (maximzasorin) [reply test/spec/css.js]: Fixed.

File: src/css.js (hunk 2 above)
Comment (maximzasorin) [reply src/css.js]: Fixed.

**Comment by maximzasorin** [comment]
> When sending the tests to BrowserStack, the CSS tests fail in IE and Edge:
> 
> Windows 8, Internet Explorer 10.0
> Windows 7, Internet Explorer 11.0
> Windows 10, Edge 17.0
> Windows 10, Edge 18.0
> Without any useful error/trace (Unspecified AssertionError). Do you happen to have one of these ready to test/debug?

I have none of this, but I debugged it in Edge on BrowserStack and found an error. It looks like Edge otherwise processes the value `undefined` for the priority argument. Edge accepts only the empty string or "important", otherwise, the value of the property does not change. And this behavior is very similar to what is described in the [specification](https://drafts.csswg.org/cssom/#dom-cssstyledeclaration-setproperty) and differs from [MDN](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration/setProperty):

> 4. If priority is not the empty string and is not an ASCII case-insensitive match for the string "important", then return.

So I replaced the default value for the important argument from `undefined` to an empty string. Now the tests are correctly executed in Edge, if you have the opportunity to test in IE, then please do it.

**Comment by maximzasorin** [comment]
@webpro 

I tested the code in IE 11 and found a problem in the work of the `setProperty` method.

If we already have a property with priority = important:

```js
element[0].style.setProperty('font-size', '15px', 'important')
```

And then try to set the same property without priority:

```js
element[0].style.setProperty('font-size', '20px', '')
```

That in Chrome, Firefox, Safari and Edge we get the value=20px and priority='', but in IE we get the value=15px and priority='important'. It turns out that the `setProperty` method in IE does not overwrite the value of the property that is set to important priority.

The only way out I see is to remove the property before setting a new value, something like is:

```js
if (!important) {
    element[0].style.removeProperty('font-size')
}
element[0].style.setProperty('font-size', '20px', '')
```

What do you think about it?

**Comment by webpro** [comment]
Apologies for my late response. First of all, thanks a lot for putting in all the efforts and test it in IE11! Overall, by now I think the required changes/maintenance to support a bad practice (usage of `!important` in CSS) do not outweigh the advantages. It's also better practice to use and (un)set classes (over `.css()` anyway, so therefore I'm going to close this issue. I hope you agree with me. Again, thanks for your time and interest in DOMtastic!

**Comment by maximzasorin** [comment]
Thanks for the comment.

**Pull request closed without merging**